File: 2696-minimum-string-length-after-removing-substrings/2696-minimum-string-length-after-removing-substrings.cpp
class Solution {
public:
    int minLength(string s) {
        int n = s.length();
        int i = 0, j = 1;
        while (j < n) {
            if (i < 0) {
                i++;
                s[i] = s[j];
            } else if ((s[i] == 'A' && s[j] == 'B') || (s[i] == 'C' && s[j] == 'D')) {
                i--;
            } else {
                i++;
                s[i] = s[j];
            }
            j++;
        }
        return i + 1;
    }
};







// class Solution {
// public:
//     int minLength(string s) {
//         stack<char> st;
//         for (char &c : s) {
//             if (st.empty()) {
//                 st.push(c);
//             } else {
//                 if (c == 'B' && st.top() == 'A') {
//                     st.pop();
//                 } else if (c == 'D' && st.top() == 'C') {
//                     st.pop();
//                 } else {
//                     st.push(c);
//                 }
//             }
//         }
//         return st.size();
//     }
// };




// class Solution {
// public:
//     int minLength(string s) {
//         while (s.find("AB") != -1 || s.find("CD") != -1) {
//             if (s.find("AB") != -1) {
//                 s.erase(s.find("AB"), 2);
//             } else if (s.find("CD") != -1) {
//                 s.erase(s.find("CD"), 2);
//             }
//         }
//         return s.length();
//     }
// };

File: LeetCode/2696-minimum-string-length-after-removing-substrings/2696-minimum-string-length-after-removing-substrings.py
class Solution:
    def minLength(self, s: str) -> int:
        n = len(s)
        s = list(s)
        i, j = 0, 1
        
        while j < n:
            if i < 0:
                i += 1
                s[i] = s[j]
            elif (s[i] == 'A' and s[j] == 'B') or (s[i] == 'C' and s[j] == 'D'):
                i -= 1
            else:
                i += 1
                s[i] = s[j]
            j += 1
        
        return i + 1 







# class Solution:
#     def minLength(self, s: str) -> int:
#         stack = []
        
#         for c in s:
#             if not stack:
#                 stack.append(c)
#             else:
#                 if c == 'B' and stack[-1] == 'A':
#                     stack.pop()
#                 elif c == 'D' and stack[-1] == 'C':
#                     stack.pop()
#                 else:
#                     stack.append(c)
        
#         return len(stack)





# class Solution:
#     def minLength(self, s: str) -> int:
#         while "AB" in s or "CD" in s:
#             if "AB" in s:
#                 s = s.replace("AB", "", 1)
#             elif "CD" in s:
#                 s = s.replace("CD", "", 1)
#         return len(s)